### ISSUE TITLE
Is there a way to list all storages/streams with full paths?

### ISSUE THREAD

**Issue by Sonic-The-Hedgehog-LNK1123** [ironfede/openmcdf#66]
I need to enumerate all the storages and streams recursively, but when using `CFStorage.VisitEntries(action, true)` I can only get the name property, but then I've no way of knowing to what storage the substorage or stream belongs to!

I want to get from this:
```
(Root)
|
+---UserForm1
|   |
|   +---o
|   |
|   +---f
|   |
|   +---CompObj
|   |
|   \---VBFrame
|
+---VBA
|   |
|   +---UserForm1
|   |
|   +---dir
|   |
|   +---ThisDocument
|   |
|   \---_VBA_PROJECT
|
+---PROJECT
|
\---PROJECTwm
```
to this:
```
UserForm1\o
UserForm1\f
UserForm1\CompObj
UserForm1\VBFrame
VBA\UserForm1
VBA\dir
VBA\ThisDocument
VBA\_VBA_PROJECT
PROJECT
PROJECTwm
```
This is probably what #63 meant as well, but that is unclear.

**Comment by ironfede** [comment]
@Sonic-The-Hedgehog-LNK1123 , currently there is no such feature. You should customize VisitEntries method to add a "path" parameter in recursive section of the visitor. Anyway this is an enhancement I will include in a next release.
If you find a solution, please feel free to open a pull request for integration.
Thank you,
Best Regards
Federico

**Comment by jtran** [comment]
Has there been any progress made on this?  It's essential for an application I'm working on to know what storage each stream originated from.

I noticed that in the Structured Storage Explorer, it does this by calling `VisitEntries(action, false)`, and then recursing inside the action.  See [MainForm.AddNodes()](https://github.com/ironfede/openmcdf/blob/HEAD/sources/Structured%20Storage%20Explorer/MainForm.cs#L192).  Is that a safe workaround?

**Comment by farfilli** [comment]
What about being able to retrieve the filename over the root? I haven't found a way to do it.

**Comment by jeremy-visionaid** [comment]
@Sonic-The-Hedgehog-LNK1123 @farfilli Better late than never... I've added a `Path` property to the `EntryInfo` in the v3.0 proof of concept:

![image](https://github.com/user-attachments/assets/e686b9f7-38fb-450c-bcc4-d46b43792a76)

Is this sufficient for your purposes?

**Comment by farfilli** [comment]
My problem was accessing the file stream name, which was a non-public property.
I have fixed it through reflections but now I see that on the V3 is public so it completely fulfills my needing

Thanks

**Comment by jeremy-visionaid** [comment]
@farfilli v3 is based around abstract streams, so it's not necessarily tied to a file. However, I've added a property to expose the base Stream and you can use pattern matching from there to get the file name if the base stream is a FileStream. Otherwise, I  think it would be up to the developer to store the filename separately.

**Comment by jeremy-visionaid** [comment]
@farfilli I'll close this one as fixed by #194 . Please go ahead and open a new issue though if the above is not sufficient for your case.
